src/main/java/org/springframework/data/repository/query/QueryMethodValueEvaluationContextAccessor.java

```diff
@@ -0,0 +1,254 @@
+
+		this.environment = environment;
+		this.evaluationContextProvider = createEvaluationContextProvider(extensions);
+	}
+
+	private static EvaluationContextProvider createEvaluationContextProvider(ListableBeanFactory factory) {
+
+		return ReactiveWrappers.isAvailable(ReactiveWrappers.ReactiveLibrary.PROJECT_REACTOR)
+				? new ReactiveExtensionAwareEvaluationContextProvider(factory)
+				: new ExtensionAwareEvaluationContextProvider(factory);
+
+	}
+
+	private static EvaluationContextProvider createEvaluationContextProvider(
+			Collection<? extends ExtensionIdAware> extensions) {
+
+		return ReactiveWrappers.isAvailable(ReactiveWrappers.ReactiveLibrary.PROJECT_REACTOR)
+				? new ReactiveExtensionAwareEvaluationContextProvider(extensions)
+				: new ExtensionAwareEvaluationContextProvider(extensions);
+	}
+
+	EvaluationContextProvider getEvaluationContextProvider() {
+		return evaluationContextProvider;
+	}
+
+	/**
+	 * Creates a new {@link ValueEvaluationContextProvider} for the given {@link Parameters}.
+	 *
+	 * @param parameters must not be {@literal null}.
+	 * @return a new {@link ValueEvaluationContextProvider} for the given {@link Parameters}.
+	 */
+	public ValueEvaluationContextProvider create(Parameters<?, ?> parameters) {
+
+		Assert.notNull(parameters, "Parameters must not be null");
+
+		if (ReactiveWrappers.isAvailable(ReactiveWrappers.ReactiveLibrary.PROJECT_REACTOR)) {
+			if (evaluationContextProvider instanceof ReactiveEvaluationContextProvider reactive) {
+				return new DefaultReactiveQueryMethodValueEvaluationContextProvider(environment, parameters, reactive);
+			}
+		}
+
+		return new DefaultQueryMethodValueEvaluationContextProvider(environment, parameters, evaluationContextProvider);
+	}
+
+	/**
+	 * Exposes variables for all named parameters for the given arguments. Also exposes non-bindable parameters under the
+	 * names of their types.
+	 *
+	 * @param parameters must not be {@literal null}.
+	 * @param arguments must not be {@literal null}.
+	 * @return
+	 */
+	static Map<String, Object> collectVariables(Parameters<?, ?> parameters, Object[] arguments) {
+
+		if (parameters.getNumberOfParameters() != arguments.length) {
+
+			throw new IllegalArgumentException(
+					"Number of method parameters (%d) must match the number of method invocation arguments (%d)"
+							.formatted(parameters.getNumberOfParameters(), arguments.length));
+		}
+
+		Map<String, Object> variables = new HashMap<>(parameters.getNumberOfParameters(), 1.0f);
+
+		parameters.stream()//
+				.filter(Parameter::isSpecialParameter)//
+				.forEach(it -> variables.put(//
+						StringUtils.uncapitalize(it.getType().getSimpleName()), //
+						arguments[it.getIndex()]));
+
+		parameters.stream()//
+				.filter(Parameter::isNamedParameter)//
+				.forEach(it -> variables.put(//
+						it.getName().orElseThrow(() -> new IllegalStateException("Should never occur")), //
+						arguments[it.getIndex()]));
```

Comment:
can't we put this in one stream processing instead of running over parameters twice?